src/renderer/mod.rs

```diff
@@ -0,0 +1,169 @@
+pub mod margin;
+
+use crate::display_list::DisplayList;
+use crate::snippet::Snippet;
+use anstyle::{AnsiColor, Effects, Style};
+use margin::Margin;
+use std::fmt::Display;
+
+#[derive(Clone)]
+pub struct Renderer {
+    anonymized_line_numbers: bool,
+    margin: Option<Margin>,
+    stylesheet: StyleSheet,
+}
+
+impl Renderer {
+    pub fn render<'a>(&'a self, snippet: Snippet<'a>) -> impl Display + 'a {
+        DisplayList::new(
+            snippet,
+            &self.stylesheet,
+            self.anonymized_line_numbers,
+            self.margin,
+        )
+    }
+
+    /// No terminal styling
+    pub const fn plain() -> Self {
+        Self {
+            anonymized_line_numbers: false,
+            margin: None,
+            stylesheet: StyleSheet::const_default(),
+        }
+    }
+
+    /// Default terminal styling
+    pub const fn styled() -> Self {
+        Self {
+            anonymized_line_numbers: false,
+            margin: None,
+            stylesheet: StyleSheet {
+                error: AnsiColor::BrightRed.on_default().effects(Effects::BOLD),
+                warning: AnsiColor::BrightYellow.on_default().effects(Effects::BOLD),
+                info: AnsiColor::BrightBlue.on_default().effects(Effects::BOLD),
+                note: Style::new().effects(Effects::BOLD),
+                help: AnsiColor::BrightCyan.on_default().effects(Effects::BOLD),
+                line_no: AnsiColor::BrightBlue.on_default().effects(Effects::BOLD),
+                emphasis: Style::new().effects(Effects::BOLD),
+                none: Style::new(),
+            },
+        }
+    }
+
+    pub const fn anonymized_line_numbers(mut self, anonymized_line_numbers: bool) -> Self {
+        self.anonymized_line_numbers = anonymized_line_numbers;
+        self
+    }
```

Comment:
`question`: Why do you need those setters, why not make the properties public?

Reply:
I did it because it is easier (and cleaner) to edit one specific field than making the fields public. The other thing is I believe it makes adding a field not a breaking change, but I do not know for certain.

Reply:
I don't understand how is it cleaner. Comparing:

```rust
let mut renderer = Renderer::plain();
renderer.anonymized_line_numbers(true);
```

vs

```rust
let mut renderer = Renderer::plain();
renderer.anonymized_line_numbers = true;
```

Can you explain further?

> The other thing is I believe it makes adding a field not a breaking change, but I do not know for certain.

That can be achieved via https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute

Reply:
That's true I forgot about that. I will switch things over

Reply:
I was looking at making them public, and I realized what I meant by cleaner.
```rust
let renderer = Renderer::plain()
    .anonymized_line_numbers(true)
    .margin(margin);
```

vs 

```rust
let mut renderer = Renderer::plain();
renderer.anonymized_line_numbers = true;
renderer.margin = margin;
```

It makes it so that `renderer` is not mutable and things can be chained together (omitting `renderer` each time).

It follows the builder pattern which I am a fan for smaller structs, for larger structs I do think making fields public is a good idea. 

If you would still like me to change it to be public fields I can but it will take me a moment to refactor.

Reply:
I wish you could use `..` initialization with `non_exhaustive`

Reply:
> It makes it so that renderer is not mutable and things can be chained together (omitting renderer each time).

That's not correct. The renderer still has to be mutable for the setters to work (you're consuming them btw. `mut self`).

The builder pattern is useful when your setters actually perform computations, invariant validation etc. In this case, your builder is the struct itself, so exposing it as such allows customers to use any Rust patterns to construct it.

By hiding it you're relying on custom uninspectable API to achieve the same.

It's a glorified version of:

```rust
struct Point {
    x: f64,
    y: f64
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Self { x, y }    
    }

    fn set_x(&mut self, x: f64) -> &mut Self {
        self.x = x;
        self
    }

    fn set_y(&mut self, y: f64) -> &mut Self {
        self.y = y;
        self
    }
}
```

vs just:

```rust
struct Point {
    pub x: f64,
    pub y: f64,
}
```

and let customers create it and operate on it any way they want. You can add helper impl methods (like you do with `plain`), but taking away ability to operate on the fields without any reason seems like a way that limits your downstream consumers from being able to interact with the struct in ways that you don't necessarily have to be able to even predict.

In other words, if at the moment of consumption, you don't care how this struct has been constructred (maybe someone created their own methods to build the Renderer, or merge two), and you don't perform any validation, or variant locking, then just let it be an open struct and let people use all of Rust expressiveness to operate on it.

Saying that - this is my opinion, not a strong requirement - I will not try to block your PR on that, so if my arguments are not convincing to you, continue with this API and I'll r+ it

Reply:
You can: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6ddedead395636105b9830d26ab0e663

Reply:
Thats within the same visibility. I don't think you can if you don't have visibility.

One benefit I've seen to builder is the ability to evolve an API, particularly for evolving styles. You'll also see builder used in a lot of the ecosystem when no validation is done.

Reply:
Ah, good call. Yeah, ability to use `..` would be nice.

I'm generally supportive of builder patterns when there's logic involved. I find it inelegant when we have a plain struct and we populate it with dummy getters/setters. 

But as I stated, this is my preference, and I'm not intending to push for it any further. I support the direction the author of the PR decides to take.

Reply:
I can see the merits of both and could go either way on this. Even though it is a breaking change, we could always change in the future. I'd say we should go with the builder (for now).